hooks/pre/31-upgrade-puppet.rb

```diff
@@ -0,0 +1,65 @@
+def puppet4_available?
+  success = []
+  success << Kafo::Helpers.execute('yum info puppet-agent')
+  success << Kafo::Helpers.execute('yum info puppetserver')
+  success.include?(false)
+end
+
+def stop_services
+  Kafo::Helpers.execute('katello-service stop')
+end
+
+def upgrade_puppet_package
+  Kafo::Helpers.execute('yum remove -y puppet-server')
+  Kafo::Helpers.execute('yum install -y puppetserver')
+  Kafo::Helpers.execute('yum install -y puppet-agent')
+end
+
+def copy_data
+  success = []
+  success << Kafo::Helpers.execute('cp -rfp /etc/puppet/environments/* /etc/puppetlabs/code/environments') if File.directory?('/etc/puppet/environments')
```

Comment:
Doesn't seem to be, I tested a p3 -> p4 upgrade that had published content and it was working.